src/TreeSitter/Importing.hs

```diff
@@ -1,116 +1,147 @@
-import TreeSitter.Tree as TS
-import qualified Data.Text as Text
-import Control.Effect
-import Control.Effect.Reader
-import Control.Monad.IO.Class
-import Data.Text.Encoding
-import qualified Data.ByteString as B
-import Control.Applicative
-import Control.Monad (void)
-
-data Expression
-      = NumberExpression Number | IdentifierExpression Identifier
-      deriving (Eq, Ord, Show)
-
-data Number = Number
-      deriving (Eq, Ord, Show)
-
-data Identifier = Identifier
-      deriving (Eq, Ord, Show)
-
-
-importByteString :: (Importing t) => Ptr TS.Parser -> ByteString -> IO (Maybe t)
-importByteString parser bytestring =
-  unsafeUseAsCStringLen bytestring $ \ (source, len) -> alloca (\ rootPtr -> do
-      let acquire =
-            ts_parser_parse_string parser nullPtr source len
-
-      let release t
-            | t == nullPtr = pure ()
-            | otherwise = ts_tree_delete t
-
-      let go treePtr =
-            if treePtr == nullPtr
-              then pure Nothing
-              else do
-                ts_tree_root_node_p treePtr rootPtr
-                withCursor (castPtr rootPtr) $ \ cursor ->
-                  Just <$> runM (runReader cursor (runReader bytestring import'))
-      Exc.bracket acquire release go)
-
-withCursor :: Ptr TSNode -> (Ptr Cursor -> IO a) -> IO a
-withCursor rootPtr action = allocaBytes sizeOfCursor $ \ cursor -> Exc.bracket_
-  (ts_tree_cursor_new_p rootPtr cursor)
-  (ts_tree_cursor_delete cursor)
-  (action cursor)
-
-instance Importing a => Importing [a] where
-  import' = push $ do
-    a <- import' @a
-    pure [a]
-
+import           GHC.Generics
+import           TreeSitter.Cursor as TS
+import           TreeSitter.Language as TS
+import           TreeSitter.Node as TS
+import           TreeSitter.Parser as TS
+import           TreeSitter.Tree as TS
+
+-- Parse source code and produce AST
+parseByteString :: Building t => Ptr TS.Language -> ByteString -> IO (Maybe t)
+parseByteString language bytestring = withParser language $ \ parser -> withParseTree parser bytestring $ \ treePtr ->
+  if treePtr == nullPtr then
+    pure Nothing
+  else
+    withRootNode treePtr $ \ rootPtr ->
+      withCursor (castPtr rootPtr) $ \ cursor ->
+        runMaybeC (runM (runReader cursor (runReader bytestring buildNode)))
+
+-- | Building is the process of iterating over tree-sitter’s parse trees using its tree cursor API and producing Haskell ASTs for the relevant nodes.
+--
+--   Datatypes which can be constructed from tree-sitter parse trees may use the default definition of 'buildNode' providing that they have a suitable 'Generic' instance.
+class Building a where
+  buildNode :: (Alternative m, Carrier sig m, Member (Reader ByteString) sig, Member (Reader (Ptr Cursor)) sig, MonadIO m) => m a
+  default buildNode :: (Alternative m, Carrier sig m, GBuilding (Rep a), Generic a, Member (Reader ByteString) sig, Member (Reader (Ptr Cursor)) sig, MonadIO m) => m a
+  buildNode = to <$> gbuildNode
```

Comment:
`gbuildNode` is what we use to match the node, and then fmapping `to :: Rep a x -> a` over it just allows us to wrap that representation into the `a` we want.